README
NAME
    Plack::Middleware::Debug::Redis - Extend Plack::Middleware::Debug with
    Redis panels

VERSION
    version 0.01

SYNOPSIS
        # inside your psgi app
        use Plack::Builder;

        my $app = sub { [ 200, [ 'Content-Type' => 'text/html' ], [ '<html><body>OK</body></html>' ] ] };
        my $redis_host = 'redi.example.com:6379';

        builder {
            mount '/' => builder {
                enable 'Debug',
                    panels => [
                        [ 'Redis::Info', server => $redis_host ],
                        [ 'Redis::Keys', server => $redis_host, db => 3 ],
                    ];
                $app;
            };
        };

DESCRIPTION
    This distribution extends Plack::Middleware::Debug with some Redis
    panels. At the moment, the following panels available:

PANELS
  Redis::Info
    Diplay panel with generic Redis server information which is available by
    the command INFO. See Plack::Middleware::Debug::Redis::Info for
    additional information.

  Redis::Keys
    Diplay panel with keys Redis server information. See
    Plack::Middleware::Debug::Redis::Keys for additional information.

BUGS
    Please report any bugs or feature requests through the web interface at
    <https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues>

SEE ALSO
    Plack::Middleware::Debug::Redis::Info

    Plack::Middleware::Debug::Redis::Keys

    Plack::Middleware::Debug

    Redis

AUTHOR
    Anton Gerasimov, <chim@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2013 by Anton Gerasimov

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Plack::Middleware::Debug::Redis.

{{$NEXT}}

0.03    Thu Jun 20 21:13:08 2013 +0000
    - set minimum perl version to 5.8.8
    - update documentation
    - tests were refactored
    - dzillify distribution

0.02  Mon Jan 14 05:35:26 2013
    - added LICENSE
    - updated build prerequisites
    - in *.t Redis mocked with Test::MockObject

0.01  Tue Jan  8 12:30:26 2013
	- original version


t/lib/FakeRedis.pm
package FakeRedis;

use strict;
use warnings;
{

    package Redis;
    $INC{'Redis.pm'} = 1;
    our $VERSION = 2;
    our $AUTOLOAD;

    my $INFO = {
        'redis_version'     => '0.1.99',
        'db0'               => 'keys=167,expires=145',
        'db1'               => 'keys=75,expires=0',
        'uptime_in_seconds' => '1591647',
        'role'              => 'master',
    };

    my $KEYS = {
        # STRING
        'coy:knows:pseudonoise:codes'       => [ 'string', 9000 ],
        # LIST
        'six:slimy:snails:sailed:silently'  => [ 'list', 35 ],
        # HASH
        'eleven:benevolent:elephants'       => [ 'hash', 17 ],
        # SET
        'two:tried:and:true:tridents'       => [ 'set', 101 ],
        # ZSET
        'tie:twine:to:three:tree:twigs'     => [ 'zset', 66 ],
    };

    sub new {
        my ( $class, %args ) = @_;
        bless \%args => $class;
    }

    sub keys {
        keys %$KEYS;
    }

    sub type {
        exists $KEYS->{$_[1]} ? $KEYS->{$_[1]}->[0] : undef;
    }

    sub info {
        $INFO;
    }

    sub _generic_len {
        exists $KEYS->{$_[1]} ? $KEYS->{$_[1]}->[1] : undef;
    }

    *strlen = *_generic_len;
    *hlen   = *_generic_len;
    *llen   = *_generic_len;
    *scard  = *_generic_len;
    *zcard  = *_generic_len;

    sub AUTOLOAD {
        shift;
        my $name = $AUTOLOAD;
        $name =~ s/.*://;
        return $name, @_;
    }
}

1;


Makefile.PL
use v5.10.1;
use ExtUtils::MakeMaker;

my $meta_merge = {
    META_MERGE => {
        resources => {
            repository => 'https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis',
            bugtracker => 'https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues',
        },
    }
};

my %options = (
    'NAME'          => 'Plack::Middleware::Debug::Redis',
    'VERSION_FROM'  => 'lib/Plack/Middleware/Debug/Redis.pm',
    'AUTHOR'        => 'Anton Gerasimov <chim@cpan.org>',
    'ABSTRACT_FROM' => 'lib/Plack/Middleware/Debug/Redis.pm',
    'LICENSE'       => 'perl',
    'PREREQ_PM'     => {
        'Redis'                    => '1.955',
        'Plack::Middleware::Debug' => 0,
        'strict'                   => 0,
    },
    'BUILD_REQUIRES' => {
        'Test::More' => '0.98',
    },
    'MIN_PERL_VERSION' => '5.10.1',
    'test' => {
        'TESTS' => 't/*.t',
    },
    'clean' => {
        'FILES' => '*.tar.gz',
    },
    $ExtUtils::MakeMaker::VERSION >= 6.50 ? (%$meta_merge) : (),
);

WriteMakefile(%options);


lib/Plack/Middleware/Debug/Redis.pm
package Plack::Middleware::Debug::Redis;

# ABSTRACT: Extend Plack::Middleware::Debug with Redis panels

use strict;
use warnings;
use Carp ();
use Redis 1.955;
use Plack::Util::Accessor qw/instance password db reconnect every debug redis/;

# VERSION
# AUTHORITY

sub redis_connect {
    my ($self, $args) = @_;

    my $croak = sub { Carp::croak $_[0] };

    my %options = (
        debug     => $self->debug     || 0,
        reconnect => $self->reconnect || 10,
        every     => $self->every     || 100,
    );

    $options{password} = $self->password if $self->password;

    my $instance = $self->_parse_instance($self->instance);

    if ($instance->{unix}) {
        $croak->("Nonexistent redis socket ($instance->{thru})!") unless -e $instance->{thru} && -S _;
    }

    $options{ $instance->{unix} ? 'sock' : 'server' } = $instance->{thru};

    $self->db($self->db || 0);

    my $_handle;
    eval { $_handle = Redis->new(%options) };
    $croak->("Cannot get redis handle: $@") if $@;

    $self->redis($_handle);
}

sub _parse_instance {
    my ($self, $instance) = @_;

    my $params = { unix => 0, thru => '127.0.0.1:6379' };

    # slightly improved piece of code from Redis.pm by Pedro Melo (cpan:MELO)
    CHANCE: {
        last CHANCE unless $instance;

        if ($instance =~ m,^(unix:)?(/.+)$,i) {
            $params->{thru} = $2;
            $params->{unix} = 1;
            last CHANCE;
        }
        if ($instance =~ m,^((tcp|inet):)?(.+)$,i) {
            my ($server, $port) = ($3, undef);
            ($server, $port)    = split /:/, $server;
            $params->{thru}     = lc($server) . ':' . (($port && ($port > 0 && $port <= 65535)) ? $port : '6379');
        }
    }

    $params;
}

1; # End of Plack::Middleware::Debug::Redis

__END__

=pod

=head1 SYNOPSIS

    # inside your psgi app
    use Plack::Builder;

    my $app = sub {[
        200,
        [ 'Content-Type' => 'text/html' ],
        [ '<html><body>OK</body></html>' ]
    ]};
    my $redis_host = 'redi.example.com:6379';

    builder {
        mount '/' => builder {
            enable 'Debug',
                panels => [
                    [ 'Redis::Info', instance => $redis_host ],
                    [ 'Redis::Keys', instance => $redis_host, db => 3 ],
                ];
            $app;
        };
    };

=head1 DESCRIPTION

This distribution extends Plack::Middleware::Debug with some Redis panels. At the moment, the following panels
available:

=head1 PANELS

=head2 Redis::Info

Diplay panel with generic Redis server information which is available by the command INFO.
See L<Plack::Middleware::Debug::Redis::Info> for additional information.

=head2 Redis::Keys

Diplay panel with keys Redis server information. See L<Plack::Middleware::Debug::Redis::Keys>
for additional information.

=head1 METHODS

=head2 redis_connect

Checks passed parameters and connects to redis server instance. Returns redis handle or croaks.

=head2 redis

Redis handle to operate with.

=head1 OPTIONS

All options should be passed as a hash reference. The following options are available to tune it for your needs.

=head2 instance

A string consist of a hostname (or an IP address) and port number (delimited with a colon) or unix socket path
of the redis-server instance to connect to. Not required. Default value is B<127.0.0.1:6379>. Some usage examples

    # tcp/ip redis-servers
    instance => '';                          # treats as '127.0.0.1:6379'
    instance => 'TCP:example.com:11230';     # ..as 'example.com:11230'
    instance => 'tcp:redis.example.org';     # ..as 'redis.example.org:6379'
    instance => 'redis-db.example.com';      # ..as 'redis-db.example.com:6379'
    instance => 'tcp:127.0.0.1';             # ..as '127.0.0.1:6379'
    instance => 'tcp:10.90.90.90:5000';      # ..as '10.90.90.90:5000'
    instance => '192.168.100.230';           # ..as '192.168.100.230:6379'
    instance => 'bogus:0'                    # ..as 'bogus:6379' (allowed > 0 and < 65536)
    instance => 'Inet:172.16.5.4:65000';     # ..as '172.16.5.4:65000'
    instance => 'bar:-100';                  # ..as 'bar:6379' (allowed > 0 and < 65536)
    instance => 'baz:70000';                 # ..as 'baz:6379' (allowed > 0 and < 65536) and so on..

    # unix sockets might be passed like this
    instance => 'Unix:/var/foo/Redis.sock';  # this socket path '/var/foo/Redis.sock'
    instance => '/bar/tmp/redis/sock';       # ..as '/bar/tmp/redis/sock',
    instance => 'unix:/var/foo/redis.sock';  # ..as '/var/foo/redis.sock',

=head2 password

Password string for redis-server's AUTH command to processing any other commands. Optional. Check the redis-server
manual for directive I<requirepass> if you would to use redis internal authentication.

=head2 db

A redis-server database number to use. Not obligatory option. If this one omitted then value B<0> will
be assigned.

=head2 reconnect

A time (in seconds) to re-establish connection to the redis-server before an exception will be raised. Not required.
Default value is B<10> sec.

=head2 every

Interval (in milliseconds) after which will be an attempt to re-establish lost connection to the redis-server. Not required.
Default value is B<100> ms.

=head2 debug

Enables debug information to STDERR, including all interactions with the redis-server. Not required.
Default value is B<0> (disabled).

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues>

=head1 SEE ALSO

L<Plack::Middleware::Debug::Redis::Info>

L<Plack::Middleware::Debug::Redis::Keys>

L<Plack::Middleware::Debug>

L<Redis>

=cut


lib/Plack/Middleware/Debug/Redis/Info.pm
package Plack::Middleware::Debug::Redis::Info;

# ABSTRACT: Redis info debug panel

use strict;
use warnings;
use parent qw(Plack::Middleware::Debug::Base Plack::Middleware::Debug::Redis);

# VERSION
# AUTHORITY

sub prepare_app {
    my ($self) = @_;

    $self->redis_connect;
}

sub run {
    my ($self, $env, $panel) = @_;

    $panel->title('Redis::Info');
    $panel->nav_title($panel->title);

    my $info = $self->redis->info;

    # tweak db keys
    foreach my $db (grep { /^db\d{1,2}/ } keys %$info) {
        my $flatten = $self->flatten_db($db, $info->{$db});
        my @keys_flatten = keys %$flatten;
        @$info{@keys_flatten} = @$flatten{@keys_flatten};
        delete $info->{$db};
    }

    $panel->nav_subtitle('Version: ' . $info->{redis_version});

    return sub {
        $panel->content($self->render_hash($info));
    };
}

sub flatten_db {
    my ($self, $database, $value) = @_;

    my %flatten = ();

    %flatten = map {
        my @ary = split /=/;
        $database . '_' . $ary[0] => $ary[1];
    } split /,/, $value;

    \%flatten;
}

1; # End of Plack::Middleware::Debug::Redis::Info

__END__

=pod

=head1 SYNOPSIS

    # inside your psgi app
    enable 'Debug',
        panels => [
            [ 'Redis::Info', instance => 'redis.example.com:6379' ],
        ];

=head1 DESCRIPTION

Plack::Middleware::Debug::Redis::Info extends Plack::Middleware::Debug by adding redis server info debug panel.
Panel displays data which available through INFO command issued in redis-cli. Before displaying info some tweaks
were processed. Normally INFO command shows total and expires keys in one line such as

    db0 => 'keys=167,expires=145',
    db1 => 'keys=75,expires=0',

This module turn in to

    db0_expires => '145',
    db0_keys    => '167',
    db1_expires => '0',
    db1_keys    => '75',

=head1 METHODS

=head2 prepare_app

See L<Plack::Middleware::Debug>

=head2 run

See L<Plack::Middleware::Debug>

=head2 redis_connect

See L<Plack::Middleware::Debug::Redis>

=head2 redis

See L<Plack::Middleware::Debug::Redis>

=head2 flatten_db

Flatten some complex data structures got from redis' INFO command. At the moment this is keys' composition of the database.

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues>

=head1 SEE ALSO

L<Plack::Middleware::Debug::Redis>

L<Plack::Middleware::Debug>

L<Redis>

=cut


lib/Plack/Middleware/Debug/Redis/Keys.pm
package Plack::Middleware::Debug::Redis::Keys;

# ABSTRACT: Redis keys debug panel

use strict;
use warnings;
use parent qw(Plack::Middleware::Debug::Base Plack::Middleware::Debug::Redis);

# VERSION
# AUTHORITY

sub prepare_app {
    my ($self) = @_;

    $self->redis_connect;
}

sub run {
    my ($self, $env, $panel) = @_;

    $panel->title('Redis::Keys');
    $panel->nav_title($panel->title);

    my %measure = (
        'HASH'   => 'hlen',
        'LIST'   => 'llen',
        'STRING' => 'strlen',
        'ZSET'   => 'zcard',
        'SET'    => 'scard',
    );

    return sub {
        my ($res) = @_;

        my ($keyz, $ktype, $klen);
        $self->redis->select($self->db);
        my @keys = $self->redis->keys('*');
        $panel->nav_subtitle('DB #' . $self->db . ' (' . scalar(@keys) . ')');

        for my $key (sort @keys) {
            $ktype = uc($self->redis->type($key));

            my $method = exists $measure{$ktype} ? $measure{$ktype} : undef;

            $klen = $method ? $self->redis->$method($key) : undef;

            $keyz->{$key} = $ktype . ($klen ? ' (' . $klen . ')' : '');
        }

        $panel->content($self->render_hash($keyz));
    };
}

1; # End of Plack::Middleware::Debug::Redis::Keys

__END__

=pod

=head1 SYNOPSIS

    # inside your psgi app
    enable 'Debug',
        panels => [
            [ 'Redis::Keys', instance => 'redis.example.com:6379', db => 3 ],
        ];

=head1 DESCRIPTION

Plack::Middleware::Debug::Redis::Keys extends Plack::Middleware::Debug by adding redis server keys debug panel.
Panel displays available keys in the redis database and its type.

    coy:knows:pseudonoise:codes             STRING (9000)
    six:slimy:snails:sailed:silently        LIST (35)
    eleven:benevolent:elephants             HASH (17)
    two:tried:and:true:tridents             SET (101)
    tie:twine:to:three:tree:twigs           ZSET (66)

Also in brackets displays key-type specific data. For I<STRING> keys it's key length in bytes; for I<HASH> - number of fields
in a hash; for I<LIST> - length of a list (number of items); for I<SET> and I<ZSET> - number of members in a set.

This panel might be added several times for different databases. Just add it again to Plack Debug panels and provide another
database number.

=head1 METHODS

=head2 prepare_app

See L<Plack::Middleware::Debug>

=head2 run

See L<Plack::Middleware::Debug>

=head2 redis_connect

See L<Plack::Middleware::Debug::Redis>

=head2 redis

See L<Plack::Middleware::Debug::Redis>

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues>

=head1 SEE ALSO

L<Plack::Middleware::Debug::Redis>

L<Plack::Middleware::Debug>

L<Redis>

=cut
